Ocean Gitlab V2 Integration

MusahMusah opened the pull request:
# Description

What - Implemented Gitlab Integration

Why - The goal is to ensure that the integration is efficient, intuitive, and user-friendly, while minimizing dependencies. This enables better control over rate limits, supports real-time events, and allows for easier maintenance and testing.

How - By building an async HTTP client using Ocean's custom implementation, integrating rate-limiting mechanisms, and streamlining the setup process to make parameters more intuitive. Webhook support and real-time event handling will enhance user interaction, and verbosity and testing ensure clarity and reliability throughout development.

## Type of change

Please leave one option from the following and delete the rest:

- [x] New Integration (non-breaking change which adds a new integration)

<h4> All tests should be run against the port production environment(using a testing org). </h4>
I implemented my test cases by simulating those from integrations/fake-integration. However, when I run the tests, I consistently encounter the following error:

`FAILED tests/test_integration.py::test_all_resync_methods - pydantic.error_wrappers.ValidationError: 1 validation error for IntegrationConfiguration
unhashable type: 'list' (type=type_error)
`

Due to this error, I have temporarily commented out my test code [here](https://github.com/port-labs/ocean/pull/1019/files#diff-943497e1611996d1bd50717ddd79696a8f9c2b8d3bf03190177d75344f0199f1). I believe the issue may stem from the Ocean framework itself, which is why I have not removed the tests entirely. The commented-out code is retained for reference and future debugging.

Note: The test was passing in a previous version of Ocean that I had cloned locally. I suspect the error, which arises from invoking Ocean's `get_raw_result_on_integration_sync_kinds` test utility function, may have been introduced in an earlier version of the framework.

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector


### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [x] Handled rate limiting
- [x] Handled pagination
- [x] Implemented the code in async
- [x] Support Multi account

## Screenshots
<img width="1495" alt="image" src="https://github.com/user-attachments/assets/fad13e97-62cd-41fe-8fb5-3548ef88aa29">
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/418342b2-6dda-4d94-a8e6-9ff7bbebf008">
<img width="1490" alt="image" src="https://github.com/user-attachments/assets/f590a9e3-93f9-43a9-9d94-0a9d4af3f267">
<img width="1510" alt="image" src="https://github.com/user-attachments/assets/3295f583-471a-4b5c-81ef-270063f2f904">



## API Documentation
Here are the GitLab REST API endpoints that I referenced for my integrations:
- [Gitlab Projects](https://docs.gitlab.com/ee/api/projects.html)
- [Gitlab Issues](https://docs.gitlab.com/ee/api/issues.html)
- [Gitlab Groups](https://docs.gitlab.com/ee/api/groups.html)
- [Gitlab Merge Requests](https://docs.gitlab.com/ee/api/merge_requests.html)
  
**WEBHOOKS**
- [Gitlab Webhook Events](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html#issue-events)

**RATE LIMITS**
- [User Rate limit](https://docs.gitlab.com/ee/administration/settings/user_and_ip_rate_limits.html)

**PAGINATION**
- [Offset Based Pagination](https://docs.gitlab.com/ee/api/rest/#pagination)

PeyGis commented on integrations/gitlab_v2/client.py:
```diff
@@ -0,0 +1,296 @@
+    "unapproval",
+]
+DELETE_WEBHOOK_EVENTS: list[str] = ["close", "merge"]
+WEBHOOK_EVENTS_TO_TRACK: dict[str, bool] = {
+    "issues_events": True,
+    "merge_requests_events": True,
+}
```
gitlab has more webhook events. please track all of them instead of these 2

MusahMusah replied:
Okay noted

PeyGis commented on integrations/gitlab_v2/client.py:
```diff
@@ -0,0 +1,296 @@
+
+class GitlabClient:
+    def __init__(self, gitlab_host: str, gitlab_token: str) -> None:
+        self.projects_url = f"{gitlab_host}/api/v4/projects"
+        self.merge_requests_url = f"{gitlab_host}/api/v4/merge_requests"
+        self.issues_url = f"{gitlab_host}/api/v4/issues"
+        self.groups_url = f"{gitlab_host}/api/v4/groups"
+
```
This is going to cause complexity in the future if we keep adding more kinds. Just have one single url and that's what you should pass to the async http client

PeyGis commented on integrations/gitlab_v2/integration.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class GitlabProjectSelector(Selector):
+    onlyGrouped: bool = Field(
+        default=True, description="Retrieve only grouped projects"
+    )
```
may I know you meant by grouped projects?

MusahMusah replied:
The intension was to give users ability to filter only projects that belongs to atleast one group or projects that have existing relationship with group(s).

PeyGis commented on integrations/gitlab_v2/main.py:
```diff
@@ -0,0 +1,157 @@
+    ISSUE = "issue"
+
+
+class WebHookEventType(StrEnum):
+    MERGE_REQUEST = "merge_request"
+    ISSUE = "issue"
```
add more webhook types

PeyGis commented on integrations/gitlab_v2/main.py:
```diff
@@ -0,0 +1,157 @@
+    logger.info("Starting musah_gitlab integration")
+    tokens = ocean.integration_config["gitlab_access_tokens"]["tokens"]
+
+    # Check if tokens is a list and filter valid tokens (strings only)
+    if isinstance(tokens, list):
+        tokens_are_valid = filter(lambda token: isinstance(token, str), tokens)
+
+        # Ensure all tokens are valid strings
+        if not all(tokens_are_valid):
+            raise InvalidTokenException("Invalid access tokens, ensure all tokens are valid strings")
+    else:
+        raise InvalidTokenException("Invalid access tokens, confirm you passed in a list of tokens")
+
```
the validation of tokens should be something that should be handled by GitLab and not us

MusahMusah replied:
The intent is to support multiple token implementation from different GitLab accounts, allowing for a seamless resync of all resources at once, rather than requiring multiple integrations. However, I can revert to supporting a single token if this approach does not align with the intended requirements. Please let me know if i should stick to just single access token for now.

MusahMusah replied:
The validation only ensures tokens are provided in below format :
`{"tokens": ["token1", "token2"]}`

PeyGis replied:
sure. But ideally, when users provide tokens, we want it to be in this format
{"access-token-1": "list of projects or sub groups it should pull data from, you can also use * so it brings all", "access-token-2": "list of projects or sub groups it should pull data from", "access-token-X": "list of xxx"}

This way, when the user installs the integration, it will fetch data from all the tokens that have been provided

PeyGis replied:
https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/gitlab/installation#tokenmapping

You can read more here

MusahMusah replied:
I see now, thanks for clarifying this.

PeyGis commented on integrations/gitlab_v2/main.py:
```diff
@@ -0,0 +1,157 @@
+    return await bootstrap_client()
+
+
+def initialize_client(gitlab_access_token: str) -> GitlabClient:
+    return GitlabClient(
+        ocean.integration_config["gitlab_host"],
+        gitlab_access_token,
+    )
```
why do you pass the access token to this function? why can't you do the same as the gitlab_host?

MusahMusah replied:
The access token is passed to the function because it can vary dynamically depending on the request, as my implementation supports multiple access tokens.  This allows the flexibility to switch between tokens as needed. In contrast, the GitLab host is a constant value that doesn’t change, so it's directly accessed from the configuration rather than passed as a parameter, making the token the only part that requires flexibility during initialization.

MusahMusah replied:
Please let me know if there is no need to support multiple access tokens

PeyGis replied:
supporting multiple tokens is one of the main focus of this assignment so yeah, good job on that. I was only concerned about the way you've structured it

PeyGis commented on integrations/gitlab_v2/main.py:
```diff
@@ -0,0 +1,157 @@
+        logger.info("Skipping webhook creation because the event listener is ONCE")
+        return
+
+    gitlab_client = initialize_client(tokens[0])
```
can you share how the token is passed to the integration?

MusahMusah replied:
The tokens can be passed as an environment variable 
`OCEAN__INTEGRATION__CONFIG__GITLAB_ACCESS_TOKENS={"tokens": ["token1", "token2"]}`

Also validated in the `spec.yaml` like so:

> configurations:
  - name: gitlabAccessTokens
    required: true
    type: object
    sensitive: true
    description: 'Gitlab access token. See the <a target="_blank" href="https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html">Gitlab Documentation</a>'

PeyGis commented on integrations/gitlab_v2/main.py:
```diff
@@ -0,0 +1,157 @@
+        logger.info(f"Webhook event '{webhook_event}' not recognized.")
+        return {"ok": True}
+
+    if webhook_event == WebHookEventType.MERGE_REQUEST:
+        payload = extract_merge_request_payload(data)
+        await ocean_action(ResourceKind.MERGE_REQUEST, [payload])
+    elif webhook_event == WebHookEventType.ISSUE:
+        payload = extract_issue_payload(data)
+        logger.info(f"Upserting issue with payload: {payload}")
+        await ocean_action(ResourceKind.ISSUE, [payload])
+    else:
+        logger.info(f"Unhandled webhook event type: {webhook_event}")
+        return {"ok": True}
```
handle the other webhook events

MusahMusah replied:
I'm now handling "push", "merge_request", and "issue" events. These are the events associated to the resources given. Kindly specify which other events you would like my implementation to handle as well. thanks so much.

PeyGis commented on integrations/gitlab_v2/main.py:
```diff
@@ -0,0 +1,157 @@
+        async for projects in client.get_projects():
+            logger.info(f"Received {kind} batch with {len(projects)} projects")
+            if config.selector.onlyGrouped:
+                projects = [project for project in projects if project.get("__group")]
```
I don't see what this filter is solving?

MusahMusah replied:
Issue has been addressed and this filter has been removed as my implementation now works with a single resync method as suggested by other reviewer.

PeyGis commented on integrations/gitlab_v2/tests/test_integration.py:
```diff
@@ -0,0 +1,88 @@
+# import os
+# from typing import Any
+# from unittest.mock import AsyncMock
+#
+# import pytest
+#
+# from client import GitlabClient
+# from port_ocean.tests.helpers import (
+#     get_raw_result_on_integration_sync_kinds,
+# )
+# from pytest_httpx import HTTPXMock
+#
+# INTEGRATION_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), "../"))
+#
+# FAKE_PROJECTS: list[dict[str, Any]]  = [
+#     {
+#         "id": 1,
+#         "name": "Test Project",
+#         "path_with_namespace": "test-namespace/test-project",
+#         "web_url": "https://gitlab.com/test-namespace/test-project",
+#         "description": "Project description",
+#     }
+# ]
+#
+# FAKE_GROUPS: list[dict[str, Any]] = [
+#     {
+#         "id": 1,
+#         "title": "Test Group",
+#         "visibility": "private",
+#         "web_url": "https://gitlab.com/test-namespace/test-group",
+#         "description": "Group description",
+#     }
```
everything here is commented. if you don't need it, delete them

MusahMusah replied:
I commented this out because:

I implemented my test cases by simulating those from integrations/fake-integration. However, when I run the tests, I consistently encounter the following error:
FAILED tests/test_integration.py::test_all_resync_methods - pydantic.error_wrappers.ValidationError: 1 validation error for IntegrationConfiguration unhashable type: 'list' (type=type_error) 

Note: The test was passing in a previous version of Ocean that I had cloned locally. I suspect the error, which arises from invoking Ocean's get_raw_result_on_integration_sync_kinds test utility function, may have been introduced in an earlier version of the framework.

PeyGis commented on integrations/gitlab_v2/utils.py:
```diff
@@ -0,0 +1,39 @@
+from loguru import logger
+
+
+def extract_merge_request_payload(data: dict[str, Any]) -> dict[str, Any]:
+    logger.info(f"Extracting merge request for project: {data['project']['id']}")
+    return {
+        "id": data["object_attributes"]["id"],
+        "title": data["object_attributes"]["title"],
+        "author": {
+            "name": data["user"]["name"],
+        },
+        "state": data["object_attributes"]["state"],
+        "created_at": data["object_attributes"]["created_at"],
+        "updated_at": data["object_attributes"]["updated_at"],
+        "web_url": data["object_attributes"]["source"]["web_url"],
+        "reviewers": data["reviewers"][0]["name"],
+        "__project": data["project"],
+    }
+
+
+def extract_issue_payload(data: dict[str, Any]) -> dict[str, Any]:
+    logger.info(f"Extracting issue for project: {data['project']['id']}")
+    return {
+        "id": data["object_attributes"]["id"],
+        "title": data["object_attributes"]["title"],
+        "link": data["object_attributes"]["url"],
+        "description": data["object_attributes"]["description"],
+        "created_at": data["object_attributes"]["created_at"],
+        "updated_at": data["object_attributes"]["updated_at"],
+        "author": {
```
this should be done using jq in the port-app-config.yaml

MusahMusah replied:
alright then.

PeyGis reviewed (changes requested):
looks good. please go over my comments

MusahMusah commented:
Hi @PeyGis

Thank you for taking the time to review my PR. I have addressed most of the issues you raised and left a few comments for clarification. I'm looking forward to your feedback on those.
